### PR TITLE
Add icon-margin layout property

### PR DESCRIPTION
**Pull request by drwestco**
Add new icon-margin layout property to symbol layer, extending the behavior of icon-padding. icon-margin supports up to 4 values, including negative values, similar to how the CSS margin property works.

## Launch Checklist

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [X] Write tests for all new functionality.
 - [X] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by HarelM** [comment]
Can you better explain the different between padding and margin?

**Comment by xabbu42** [comment]
Reading the code padding and margin are the same and the name change is for backward compatibility. correct?

This seems a poor choice to me as this names are used for different concepts in html and that will confuse people. Alternatives would be icon-padding-{left,right,top,bottom} and/or icon-padding-perside or icon-padding-array.

Also note that it is currently very hard and sometimes impossible to to data-driven styling with arrays. you cant for example interpolate the values of an array independently, while you could do that with different properties.

**Comment by drwestco** [comment]
Yeah, I couldn't find an easy way to use the same icon-padding property and have it support the existing single-number value while also allowing an array. I picked 'margin' since CSS padding doesn't support negative values, though that's an arbitrary reason. Happy to change if someone has a better name. icon-padding-array isn't bad but there may be better options.

I did discover the limitation on data-driven styling. There's an issue in the mapbox repo asking for the ability to have an expression as an array element. It was opened in 2018 and has had no progress since.

Four separate properties for per-side padding seems like overkill.

**Comment by wipfli** [comment]
Please add a changelog item and open an issue in the native repo that this feature might be introduced in the style spec.

**Review by wipfli — COMMENTED** [review]
Thanks @drwestco for this pull request, I think it is a good idea to make the collision bounding boxes more flexible. Regarding the naming, I would prefer if we could still call it icon-padding because then it is more aligned with the existing icon-padding option. The way I would go forward is like this:
* `icon-padding` 1 value for all 4 sides, will be overridden by `icon-padding-array` if present.
* `icon-padding-array` 2, 3, or 4 values, overrides `icon-padding`.

Question: should we consider introducing a new property which can override an existing property as a breaking change?

**Comment by wipfli** [comment]
`text-color` is a property which is overridable. This was introduced in https://github.com/maplibre/maplibre-gl-js/commit/b4fd6ee030e33d7ffc612c668a61b85171fa042b Maybe you can get some inspiration from how it was handled there?

**Comment by wipfli** [comment]
@drwestco you mentioned on slack that only one type is allowed for a style property, like for example either `number` or `array`. Which challenges would you expect if we were to support multiple types and different behavior for different types? 

I think ideally we would support `icon-padding: 5` and simultaneously stuff like `icon-padding: [1, 2, 3, 4]`.

File: src/style-spec/reference/v8.json [hunk 1]
```diff
@@ -1465,6 +1465,36 @@
...
+        }
+      },
+      "expression": {
+        "interpolated": true,
+        "parameters": [
+          "zoom",
+          "feature"
+        ]
+      },
+      "property-type": "data-driven"
```
[the remark this answers is not in the document]
Comment (drwestco) [reply src/style-spec/reference/v8.json:1496]: This old issue in the MapBox repo explains it. Basically there's been a long-standing request to allow an expression as an element of an array. The case here might be the desire to have padding-top vary with zoom level. The syntax to do so is awkward when the expression has to be top-most and output the entire array. https://github.com/mapbox/mapbox-gl-js/issues/6155

**Comment by drwestco** [comment]
> @drwestco you mentioned on slack that only one type is allowed for a style property, like for example either `number` or `array`. Which challenges would you expect if we were to support multiple types and different behavior for different types?
> 
> I think ideally we would support `icon-padding: 5` and simultaneously stuff like `icon-padding: [1, 2, 3, 4]`.

The main challenge is figuring out the style syntax. One easy approach would be to change the type for `icon-padding` from `number` to `Array<number>`, and then update the style validation / reading code to treat a bare `number` the same as a one-element array. So you could then do `icon-padding: 5`, or `icon-padding: [5]`, etc.

**Comment by wipfli** [comment]
That sounds like an interesting idea @drwestco. This would the also not be a breaking change, but more like an extension I guess.

**Comment by drwestco** [comment]
Closing this and replacing with #1289

**Pull request closed without merging**